### PR TITLE
[feature] new semnidaq driver: control a SEM via the NI-DAQ API

### PR DESCRIPTION
**Pull request by pieleric**
Depends on the "nidaqmx-python" Python wrapper for using the NI-DAQmx API.
For now tested only on a NI PCIe-6361... and also known to not work
yet on older devices like the PCIe-6251.

It supports all features needed for the SECOM and SPARC, with the
exception of the counter detector. Tested succesfully on a real SEM.

It is possible to test it on a computer without an actual card by using
the NI-DAQmx with a simulated DAQ card. Instructions are in the top
comment of semnidaq.py .

Also introduces two microscope files: one to use a real DAQ board, only,
and one to simulate a SPARC using a simulated NI DAQ board.A

### CODE REVIEW

**Review by FCanberk — COMMENTED** [review]
i am doing the new linter's job for this PR: 

please don't forget to add the empty new lines at the end of files  :grin:

**Review by tepals — CHANGES_REQUESTED** [review]
I tried to look into everything as much as possible, but I'm sure I missed something. Let me know if there is a specific part or function that you want me to look into in more detail.

File: src/odemis/driver/semnidaq.py [hunk 1]
```diff
@@ -0,0 +1,2799 @@
...
+
+    def __init__(self, name: str, role: str, children: Dict[str, Dict],
+                 device: str,
+                 multi_detector_min_period: Optional[float] = None,
+                 daemon=None, **kwargs):
+        """
+        :param children: (internal role -> kwargs) parameters setting for the children.
+            Known children are "scanner", "detector0", "detector1"...
+            They will be provided back in the .children roattribute
+        :param device: name of the NI DAQ device (ex: "Dev1"). See "nilsdev".
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/semnidaq.py:121]: It's a command. I've updated the comment.

File: src/odemis/driver/semnidaq.py [hunk 2]
```diff
@@ -0,0 +1,2799 @@
...
+
+        # we will fill the set of children with Components later in ._children
+        model.HwComponent.__init__(self, name, role, daemon=daemon, **kwargs)
+
+        # TODO: report a nice error if the nidriver is not installed (ie only nidaqmx python)
+        system = nidaqmx.system.System.local()
+
+        try:
+            self._nidev = system.devices[device]  # This accepts anything. It will fail on first call
+            self._hwVersion = f"NI {self._nidev.product_type} s/n: {self._nidev.serial_num}"  # Will fail here
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/semnidaq.py:142]: Indeed... not very clear. It fails *if it's incorrect* only on line 142. Updated the comment.

File: src/odemis/driver/semnidaq.py [hunk 3]
```diff
@@ -0,0 +1,2799 @@
...
+        """
+        full_res = self._shape[:2]
+        # adapt limits according to the scale and translation so that if scale
+        # == 1,1 and translation == 0,0 , the area is centered and a pixel is
+        # the size of pixelSize
+        roi_limits = []  # min/max for X/Y in V
+        roi_limits_raw = []  # min/max for X/Y in raw value (int16)
+        for i, lim in enumerate(self._limits):
+            center = (lim[0] + lim[1]) / 2
+            width = lim[1] - lim[0]
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/semnidaq.py:2434]: Not necessarily. But it works in both directions. I think the word is badly picked though, as it implies a positive value. => Changed to `volt_diff`.

File: src/odemis/driver/semnidaq.py [hunk 4]
```diff
@@ -0,0 +1,2799 @@
...
+            model.MD_SW_VERSION: self._swVersion,
+            model.MD_HW_VERSION: self._hwVersion,
+        }
+
+        if multi_detector_min_period is None:
+            # Use the very minimum period of the board if not specified
+            # Note that it may still give warnings
+            self._multi_detector_min_period = 1 / self._nidev.ai_max_multi_chan_rate
+        elif not isinstance(multi_detector_min_period, (float, int)):
+            raise ValueError(f"multi_detector_min_period must be float, but is {type(multi_detector_min_period)}")
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/semnidaq.py:162]: I've adjusted it to just say "a number".

File: install/linux/usr/share/odemis/sim/sparc2-nidaq-sim.odm.yaml [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+               "Spec Filter Wheel", "Spec CCD Flipper", "Spectrograph focus",
+               "Mirror Actuators", "Mirror Actuators in XY referential",
+               "Slit", "CL Selector", "CL Filter Wheel",
+               "Lens1 Mover", "Lens2 Switch"],
+}
+
+# Light (lamp with known spectrum)
+"Calibration Light": {
+    class: light.Light,
+    role: "brightlight",
```
[the remark this answers is not in the document]
Comment (pieleric) [reply install/linux/usr/share/odemis/sim/sparc2-nidaq-sim.odm.yaml]: With and without quotes are equivalent, as in YAML, strings are automatically detected. It's the same for the classes or the component names. However strings with spaces must be quoted. So we have the habit to always quote the component names. For classes and role, they are not supposed to contain spaces, so it's safe to never quote them. I've cleaned up this microscope file to drop all the quotes from the "role".

**Review by tepals — APPROVED** [review]

**Pull request merged**